src/Microsoft.OData.Core/Json/JsonReaderExtensions.cs

```diff
@@ -391,8 +392,453 @@ internal static JsonNodeType ReadNext(this IJsonReader jsonReader)
+        /// </summary>
+        /// <param name="jsonReader">The reader to inspect.</param>
+        /// <returns>A task that represents the asynchronous read operation.
+        /// The value of the TResult parameter contains the <see cref="ODataUntypedValue"/> value read from the reader.</returns>
+        internal static async Task<ODataValue> ReadAsUntypedOrNullValueAsync(this IJsonReaderAsync jsonReader)
+        {
+            StringBuilder builder = new StringBuilder();
+            await jsonReader.SkipValueAsync(builder)
+                .ConfigureAwait(false);
+            Debug.Assert(builder.Length > 0, "builder.Length > 0");
+
+            return new ODataUntypedValue()
+            {
+                RawValue = builder.ToString(),
+            };
+        }
+
+        /// <summary>
+        /// Asynchronously reads the next node from the <paramref name="jsonReader"/> as an <see cref="ODataValue"/>.
+        /// </summary>
+        /// <param name="jsonReader">The reader to inspect.</param>
+        /// <returns>A task that represents the asynchronous read operation.
+        /// The value of the TResult parameter contains the <see cref="ODataValue"/> value read from the reader.</returns>
+        internal static async Task<ODataValue> ReadODataValueAsync(this IJsonReaderAsync jsonReader)
+        {
+            if (jsonReader.NodeType == JsonNodeType.PrimitiveValue)
+            {
+                object primitiveValue = await jsonReader.ReadPrimitiveValueAsync()
+                    .ConfigureAwait(false);
+
+                return primitiveValue.ToODataValue();
+            }
+            else if (jsonReader.NodeType == JsonNodeType.StartObject)
+            {
+                await jsonReader.ReadStartObjectAsync()
+                    .ConfigureAwait(false);
+                ODataResourceValue resourceValue = new ODataResourceValue();
+                List<ODataProperty> properties = new List<ODataProperty>();
+
+                while (jsonReader.NodeType != JsonNodeType.EndObject)
+                {
+                    ODataProperty property = new ODataProperty();
+                    property.Name = await jsonReader.ReadPropertyNameAsync()
+                        .ConfigureAwait(false);
+                    property.Value = await jsonReader.ReadODataValueAsync()
+                        .ConfigureAwait(false);
+                    properties.Add(property);
+                }
+
+                resourceValue.Properties = properties;
+
+                await jsonReader.ReadEndObjectAsync()
+                    .ConfigureAwait(false);
+
+                return resourceValue;
+            }
+            else if (jsonReader.NodeType == JsonNodeType.StartArray)
+            {
+                await jsonReader.ReadStartArrayAsync()
+                    .ConfigureAwait(false);
+                ODataCollectionValue collectionValue = new ODataCollectionValue();
+                IList<object> properties = new List<object>();

```

Comment:
Using IList seems unnecessary
```suggestion
                List<object> properties = new List<object>();

```